src/org/labkey/test/tests/SecurityApiTest.java

```diff
@@ -102,30 +102,34 @@ protected void doCleanup(boolean afterTest) throws TestTimeoutException
+     * This is "one of those tests" that are seldom changed and very rarely fail. However, when it does fail you can spend
+     * more time than you want just trying to get it to run. Hopefully the following comments will help.
      *
      * If you cannot get this test to run in IntelliJ try running it from the command line. Running in IntelliJ can throw
      * a ClassNotFoundException. You are welcome to try and track it down and correct it if you want, but running from the
      * command line (gradlew uiTests) does appear to be more reliable.
      *
-     * The failed output from the test, especially on TeamCity, can be difficult to read. It is kind of a diff but not
-     * really. However, fear not! When you run the test locally from the command line you should see a line at the end of
-     * the output that looks like this:
+     * The failed output from the test can be difficult to read. Especially if the failure was between expected and actual
+     * roles and permissions. If that is the type of failure you are seeing the failure message is a rough diff between
+     * the expected JSON and the JSON that was returned. If you run the test locally from the command line you might be
+     * better able to understand what the differences are. After running locally you should see a message at the end of
+     * the run output that looks like this:
      *
      * There were failing tests. See the report at: file:///Users/janedoe/labkey/trunk/build/modules/testAutomation/test/logs/reports/html/index.html
      *
      * Open this file in a browser, click on the test name, then click on the "Standard output" button. This presents the
      * output in a more readable way (basically a log). Near the top is an "Expected:" comment with the expected output.
-     * A little ways down is a comment "Actual:" with the actual output (go figure).
-     * Looking at the expected vs. actual output in this log might be easier to compare and see what the difference is.
+     * A little ways down is a comment "Actual:" with the actual output (go figure). It might be easier to compare the
+     * expected vs. actual results from this log output.
      *
-     * The expected output comes from the security-api.xml file. Odds are you are going to have to change that file in
-     * some way. You didn't hear it from me but the easiest fix could be to replace the contents of the security-api.xml
-     * file with the actual output from the log. Obviously the smaller the change the better, but as a last resort you
-     * could try that as a fix.
+     * The expected output comes from the security-api.xml file. If this is a role and/or permission error odds are you
+     * will need to make a change to that file. Try to make the smallest change possible. If you are changing permissions,
+     * either adding or removing, you will need to update the value(s) in the effectivePermissions collection for a given role.
```

Comment:
Perhaps add a note here that the test, despite indications from the diff output, the order of the strings in the effectivePermissions collection does not matter.

Reply:
Good call. I added a comment in the paragraph above. That paragraph now reads as:

The failed output from the test can be difficult to read. Especially if the failure was between expected and actual
roles and permissions. If that is the type of failure you are seeing the failure message is a rough diff between
the expected JSON and the JSON that was returned. Unfortunately the code that checks for equality and the code
that generates the diff-error output are two different functions. The comparison code flags the two JSON objects as
not being equal. That failed check then calls the error output code which converts the JSON objects to strings. 
It compares these two string line by line, and shows all the differences, including differences in order, 
WHICH DOESN'T MATTER. This adds a lot of "noise" to the error output in the log that makes it hard to find the 
actual inconsistency you are looking for.

Reply:
You're writing a novel here. :-)